alibi_detect/cd/base.py

```diff
@@ -0,0 +1,215 @@
+            for reservoir sampling {'reservoir_sampling': n} is passed.
+        preprocess_fn
+            Function to preprocess the data before computing the data drift metrics.
+            Typically a dimensionality reduction technique.
+        preprocess_kwargs
+            Kwargs for `preprocess_fn`.
+        correction
+            Correction type for multivariate data. Either 'bonferroni' or 'fdr' (False Discovery Rate).
+        n_features
+            Number of features used in the statistical test. No need to pass it if no preprocessing takes place.
+            In case of a preprocessing step, this can also be inferred automatically but could be more
+            expensive to compute.
+        n_infer
+            Number of instances used to infer number of features from.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__()
+
+        if p_val is None:
+            logger.warning('No p-value set for the drift threshold. Need to set it to detect data drift.')
+
+        if isinstance(preprocess_kwargs, dict):  # type: ignore
+            if not isinstance(preprocess_fn, Callable):  # type: ignore
+                preprocess_fn = preprocess_drift
+            self.preprocess_fn = partial(
+                preprocess_fn,
+                **preprocess_kwargs
+            )
+            keys = list(preprocess_kwargs.keys())
+        else:
+            self.preprocess_fn, keys = None, []
+
+        # optionally already preprocess reference data
+        self.preprocess_X_ref = preprocess_X_ref
+        if preprocess_X_ref and isinstance(self.preprocess_fn, Callable):  # type: ignore
+            self.X_ref = self.preprocess_fn(X_ref)
+        else:
+            self.X_ref = X_ref
+        self.update_X_ref = update_X_ref
+        self.n = X_ref.shape[0]  # type: ignore
+        self.p_val = p_val
+        self.correction = correction
+
+        if isinstance(input_shape, tuple):
+            self.input_shape = input_shape
+        elif 'max_len' in keys:
+            self.input_shape = (preprocess_kwargs['max_len'],)
+        elif isinstance(X_ref, np.ndarray):
+            self.input_shape = X_ref.shape[1:]
+
+        # compute number of features for the univariate tests
+        if isinstance(n_features, int):
+            self.n_features = n_features
+        elif not isinstance(preprocess_fn, Callable):
+            self.n_features = X_ref.reshape(X_ref.shape[0], -1).shape[-1]
+        elif preprocess_X_ref:  # infer features from preprocessed reference data
+            self.n_features = self.X_ref.reshape(self.X_ref.shape[0], -1).shape[-1]
+        else:  # infer number of features after applying preprocessing step
+            X = self.preprocess_fn(X_ref[0:min(X_ref.shape[0], n_infer)])
+            self.n_features = X.reshape(X.shape[0], -1).shape[-1]
```

Comment:
I think this also can use `self.X_ref` as the preprocessing would already be done in that attribute?

Reply:
Ah I see in this case `X_ref` is not yet pre-processed.